### ISSUE TITLE
[apex] Prevent classes from shadowing System Namespace

### ISSUE THREAD

**Issue by noerremark** [pmd/pmd#3184]
We brainstormed about this rule in #2322. Believe it or not, I had the issue the other day, in a real environment. [Salesforce Developers Blog has a 2012 post about it](https://developer.salesforce.com/blogs/developer-relations/2012/08/default-namespaces-in-apex.html), so it's an official "feature", but pretty bad design.

**Proposed Rule Name:** ~~ClassShadowsSystemNamespace~~ TypeShadowsBuiltInNamespace

**Proposed Category:** ~~Code Style~~ Error Prone

**Description:**

**Code Sample:** 

```
public class Database {
    public static String query() {
        return 'wherefore art thou namespace?';
    }
}
```

Since `Database` is also in `System` this is a violation.

**Comment by noerremark** [comment]
@rsoesemann / @daveespo - I'll see if I can provide a PR for this.

**Comment by rsoesemann** [comment]
@noerremark please think about your implementation options. This could be a super lean XPath rule just checking a few classes against the class name. The footprint of a real Java rule is only of use when you make it work with previously unknown System classes.

**Comment by noerremark** [comment]
@rsoesemann - I agree, XPath is the best option here.

**Comment by noerremark** [comment]
@rsoesemann - I started looking into this, and there is a LOT of classes. I know XPath is the easy route, but what about doing a Java-rule that uses reflections to generate a list of bad names to test against? @adangel - I see you have been doing the stuff on jorle, would it be possible?

**Comment by rsoesemann** [comment]
@noerremark I absolutely agree that a Java rule is smart if AND ONLY if it really not only enumerates such classes but actually finds them out magically. Meaning if Salesforce adds a NEW system class this rule should auto-detect it.
If it is just a Java rule enumerating all classes I think XPath is better.

Can we reference a separate file in XPath?!

**Comment by mitchspano** [comment]
@rsoesemann 
Unfortunately, no. **Standard XPath within PMD does not have the capability to directly read data from external files.** 

While direct file access isn't possible, we can achieve similar results using these approaches:

1. **Rule Properties:** Define a rule property that allows users to specify a comma-separated list of banned class names. Our XPath expression can then use this property to check for the presence of these names.

   ```xml
   <rule ...>
       <properties>
           <property name="bannedClassNames" 
                     description="Comma-separated list of banned class names" 
                     value="foo,bar,baz" />
           <property name="xpath">
               <value>
                   <![CDATA[
                   //Identifier[contains(., '${bannedClassNames}')]
                   ]]>
               </value>
           </property>
       </properties>
   </rule>
   ```

2. **Custom Rule Implementation:**  Instead of relying solely on XPath, create a custom Java rule that extends `AbstractApexRule`. In the Java code, we can read the banned class names from a file and then analyze the AST to identify violations.

   ```java
   public class ShadowSystemNamespaceRule extends AbstractApexRule {
       private Set<String> bannedClassNames;

       public ShadowSystemNamespaceRule() {
           // Read banned words from a file
           bannedClassNames = readFromFile("system_class_names.txt"); 
       }

       @Override
       public Object visit(ASTUserClass node, Object data) {
           String identifierName = node.getImage().toLowerCase();
           if (bannedClassNames.contains(identifierName)) {
               addViolation(data, node);
           }
           return data;
       }

       // ... (Implementation for reading from file)
   }
   ```

It's also worth noting that inclusion of a supplementary configuration file in PMD's  source might change the Maven build process, and it may not be worth it. In which case, we could just have a static set of hard coded class names to watch out for, which then nudges us back in the direction of an Xpath rule.

**Comment by jsotuyod** [comment]
Since `System` classes are "auto-imported" we can't tell if a class is part of it or not if we don't add that information somewhere ourselves. It will never be "picked up automatically".

A separate file is a really bad idea for this…
1. it's non-standard within PMD
2. it needs to sit outside of the packages, or it can't be edited (beats the point)
3. users would still need to know the implementation details to edit the file to add more

I strongly believe a property (regardless of how the rule is implemented) would be the best choice.

Having said that, the rule itself seems simple enough to be implemented in XPath.

A full list of (current) `System` classes can be found [here](https://developer.salesforce.com/docs/atlas.en-us.apexref.meta/apexref/apex_namespace_System.htm)
